pkg/api/proxy.go

```diff
@@ -0,0 +1,160 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/dchest/siphash"
+	"github.com/gorilla/mux"
+
+	"zotregistry.dev/zot/pkg/api/constants"
+)
+
+func ClusterProxy(ctrlr *Controller) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			config := ctrlr.Config
+			logger := ctrlr.Log
+
+			// if no cluster or single-node cluster, handle locally
+			if config.Cluster == nil || len(config.Cluster.Members) == 1 {
+				next.ServeHTTP(response, request)
+
+				return
+			}
+
+			vars := mux.Vars(request)
+			name, ok := vars["name"]
+
+			if !ok || name == "" {
+				response.WriteHeader(http.StatusNotFound)
+
+				return
+			}
+
+			h := siphash.New([]byte(config.Cluster.HashKey))
```

Comment:
Do we always have to do a "New()" here? Why not stash it and do a `if h { h.Reset()}`

Reply:
because New() could allocate and be expensive?

Reply:
Ran a quick check on the siphash repo to compare the approaches:

Added the following code to `siphash_test.go`:
```go
func BenchmarkSum64WithLongStringInput(b *testing.B) {
	h := New([]byte("loremipsumdolors"))
	h.Write([]byte("thisisaveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylongreponame"))
	h.Sum64()
}

func BenchmarkHashWithLongStringInput(b *testing.B) {
	Hash(1234567890, 123143930393, []byte("thisisaveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylongreponame"))
}
```
Running these gave the following results:
```
➜  siphash git:(master) ✗ go test -bench '^BenchmarkSum64WithLongStringInput$'
goos: linux
goarch: amd64
pkg: github.com/dchest/siphash
cpu: Intel Core i7 9xx (Nehalem Class Core i7)
BenchmarkSum64WithLongStringInput-6   	1000000000	         0.0000013 ns/op
PASS
ok  	github.com/dchest/siphash	0.014s
➜  siphash git:(master) ✗ go test -bench '^BenchmarkHashWithLongStringInput$'
goos: linux
goarch: amd64
pkg: github.com/dchest/siphash
cpu: Intel Core i7 9xx (Nehalem Class Core i7)
BenchmarkHashWithLongStringInput-6   	1000000000	         0.0000003 ns/op
PASS
ok  	github.com/dchest/siphash	0.015s
➜  siphash git:(master) ✗ 
```
The results were repeatable in multiple executions.

The results seem to indicate that for zot's use case, there doesn't seem to be a significant change in execution time (only on the order of about 10ns/op for very large repo names). We can probably stick to the implementation that accepts a byte so that users can supply strings in the config file as opposed to accepting a string and then converting to uint64 for the Hash function (the `newDigest` method called by `New` internally does this).